src/proof_rules/mod.rs

```diff
@@ -0,0 +1,282 @@
+        let mut encoding: Option<(Rc<dyn Encoding>, Span)> = None;
+
+        let mut is_one_loop = true;
+        for s in &mut *stmts {
+            match &mut s.node {
+                StmtKind::Annotation(ident, _, _) => {
+                    if let DeclKind::AnnotationDecl(AnnotationKind::Encoding(anno_ref)) =
+                        self.tcx.get(*ident).unwrap().as_ref()
+                    {
+                        encoding = Some((anno_ref.clone(), s.span));
+                    }
+                }
+                StmtKind::Var(_) => {}
+                _ => {
+                    is_one_loop = false;
+                }
+            }
+        }
+        // If the annotated encoding only supports one loop programs throw an error
+        if let Some((enc, span)) = encoding {
+            if enc.is_one_loop() && !is_one_loop {
+                return Err(AnnotationError::OneLoopOnly(span, enc.name()));
+            }
+        }
+
+        for s in stmts {
+            self.visit_stmt(s)?;
+        }
+        Ok(())
+    }
+
+    fn visit_stmt(&mut self, s: &mut Stmt) -> Result<(), Self::Err> {
+        walk_stmt(self, s)?;
+        if let StmtKind::Annotation(ident, inputs, inner_stmt) = &mut s.node {
+            if let DeclKind::AnnotationDecl(AnnotationKind::Encoding(anno_ref)) =
+                self.tcx.get(*ident).unwrap().as_ref()
+            {
+                if let StmtKind::While(_, _) = inner_stmt.node {
+                } else {
+                    return Err(AnnotationError::NotOnWhile(s.span, *ident, s.clone()));
+                }
+
+                let mut enc_gen = anno_ref.transform(
+                    self.tcx,
+                    s.span,
+                    inputs,
+                    inner_stmt,
+                    self.direction
+                        .ok_or(AnnotationError::NotInProcedure(s.span, *ident))?,
+                )?;
+
+                let stmts = &mut enc_gen.stmts;
+                let span = enc_gen.span;
+                let decls_opt = &mut enc_gen.decls;
+
+                // Visit generated statements
+                self.visit_stmts(stmts)?;
+                s.span = span;
+                s.node = StmtKind::Block(stmts.to_vec());
+
+                if let Some(ref mut decls) = decls_opt {
+                    // Visit generated declarations
+                    self.visit_decls(decls)?;
+
+                    // Wrap generated declarations in items
+                    let items: Vec<Item<SourceUnit>> = decls
+                        .iter_mut()
+                        .map(|decl| {
+                            SourceUnit::Decl(decl.to_owned()).wrap_item(&SourceFilePath::Generated)
+                        })
+                        .collect();
+
+                    self.source_units_buf.extend(items)
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub fn walk_stmt<V: VisitorMut>(visitor: &mut V, s: &mut Stmt) -> Result<(), V::Err> {
```

Comment:
This looks weird. Shouldn't it be in the `visit.rs` file over in the `ast` module?